src/util/symbol_table.cpp

```diff
@@ -114,3 +115,97 @@ void symbol_tablet::erase(const symbolst::const_iterator &entry)
 
   internal_symbols.erase(entry);
 }
+
+/// Check whether the symbol table is in a valid state
+/// \param vm Determine whether to throw exceptions or trigger INVARIANT
+///   when validation fails
+void symbol_tablet::validate(const validation_modet vm) const
+{
+  // Check that identifiers are mapped to the correct symbol
+  for(const auto &elem : symbols)
+  {
+    const auto symbol_key = elem.first;
+    const auto &symbol = elem.second;
+
+    // Check that symbols[id].name == id
+    DATA_CHECK_WITH_DIAGNOSTICS(
+      symbol.name == symbol_key,
+      "Symbol table entry must map to a symbol with the correct identifier",
+      "Symbol table key '",
+      symbol_key,
+      "' maps to symbol '",
+      symbol.name,
+      "'");
+
+
+    // Check that the symbol basename is mapped to its full name
+    if(!symbol.base_name.empty())
+    {
+      auto base_map_search = symbol_base_map.find(symbol.base_name);
+      bool map_matches = false;
+      while (!map_matches && base_map_search != symbol_base_map.end())
+      {
+        if(base_map_search->second == symbol_key)
+          map_matches = true;
+
+        ++base_map_search;
+      }
```

Comment:
```    
      bool map_matches =
        std::find_if(
          symbol_base_map.begin(),
          symbol_base_map.end(),
          [&](const typename symbol_base_mapt::value_type &pair) {
            return pair.first == symbol.base_name && pair.second == symbol_key;
          }) != symbol_base_map.end();
```
Doesn't the while loop actually go through all the elements past the first `find` match?

Reply:
The `!map_matches` condition in the while loop will terminate the loop at the first match, but I agree that your suggested use of the std library is a better way to go - I'll update, thanks :-)

Reply:
Actually, using `equal_range` is probably better....

Reply:
Yes `equal_range` would be logarithmic, unlike the above.